File: base/src/main/java/vjson/simple/SimpleObject.java
/*
 * The MIT License
 *
 * Copyright 2019 wkgcass (https://github.com/wkgcass)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package vjson.simple;

import vjson.JSON;
import vjson.Stringifier;

import java.util.*;

public class SimpleObject extends AbstractSimpleInstance<LinkedHashMap<String, Object>> implements JSON.Object {
    private final List<SimpleObjectEntry<JSON.Instance>> map;
    private LinkedHashSet<String> keySet;
    private List<String> keyList;
    private Map<String, JSON.Instance> fastSingleMap;
    private Map<String, List<JSON.Instance>> fastMultiMap;

    public SimpleObject(Map<String, JSON.Instance> initMap) throws NullPointerException, IllegalArgumentException {
        if (initMap == null) {
            throw new NullPointerException();
        }
        for (Map.Entry<String, JSON.Instance> entry : initMap.entrySet()) {
            if (entry.getKey() == null) {
                throw new IllegalArgumentException("key should not be null");
            }
            if (entry.getValue() == null) {
                throw new IllegalArgumentException("value should not be null");
            }
        }
        this.map = new ArrayList<>(initMap.size());
        for (Map.Entry<String, JSON.Instance> entry : initMap.entrySet()) {
            this.map.add(new SimpleObjectEntry<>(entry.getKey(), entry.getValue()));
        }
    }

    public SimpleObject(List<SimpleObjectEntry<JSON.Instance>> initMap) throws NullPointerException, IllegalArgumentException {
        if (initMap == null) {
            throw new NullPointerException();
        }
        for (SimpleObjectEntry<JSON.Instance> entry : initMap) {
            if (entry == null) {
                throw new IllegalArgumentException("entry should not be null");
            }
            if (entry.key == null) {
                throw new IllegalArgumentException("key should not be null");
            }
            if (entry.value == null) {
                throw new IllegalArgumentException("value should not be null");
            }
        }
        this.map = new ArrayList<>(initMap);
    }

    protected SimpleObject(List<SimpleObjectEntry<JSON.Instance>> initMap, vjson.parser.TrustedFlag flag) {
        if (flag == null) {
            throw new UnsupportedOperationException();
        }
        this.map = initMap;
    }

    protected SimpleObject(List<SimpleObjectEntry<JSON.Instance>> initMap, vjson.util.TrustedFlag flag) {
        if (flag == null) {
            throw new UnsupportedOperationException();
        }
        this.map = initMap;
    }

    @Override
    public LinkedHashMap<String, Object> toJavaObject() {
        return new LinkedHashMap<>(super.toJavaObject());
    }

    @Override
    protected LinkedHashMap<String, Object> _toJavaObject() {
        LinkedHashMap<String, Object> javaObject = new LinkedHashMap<>();
        for (SimpleObjectEntry<JSON.Instance> entry : map) {
            javaObject.put(entry.key, entry.value.toJavaObject());
        }
        return javaObject;
    }

    @Override
    public void stringify(StringBuilder sb, Stringifier sfr) {
        sfr.beforeObjectBegin(sb, this);
        sb.append("{");
        sfr.afterObjectBegin(sb, this);
        boolean isFirst = true;
        for (SimpleObjectEntry<JSON.Instance> entry : map) {
            if (isFirst) isFirst = false;
            else {
                sfr.beforeObjectComma(sb, this);
                sb.append(",");
                sfr.afterObjectComma(sb, this);
            }
            sfr.beforeObjectKey(sb, this, entry.key);
            sb.append(JSON.String.stringify(entry.key));
            sfr.afterObjectKey(sb, this, entry.key);
            sfr.beforeObjectColon(sb, this);
            sb.append(":");
            sfr.afterObjectColon(sb, this);
            sfr.beforeObjectValue(sb, this, entry.key, entry.value);
            entry.value.stringify(sb, sfr);
            sfr.afterObjectValue(sb, this, entry.key, entry.value);
        }
        sfr.beforeObjectEnd(sb, this);
        sb.append("}");
        sfr.afterObjectEnd(sb, this);
    }

    @Override
    protected String _toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Object{");
        boolean isFirst = true;
        for (SimpleObjectEntry<JSON.Instance> entry : map) {
            if (isFirst) isFirst = false;
            else sb.append(", ");
            sb.append(entry.key).append(":").append(entry.value);
        }
        sb.append("}");
        return sb.toString();
    }

    private LinkedHashSet<String> _keySet() {
        if (this.keySet == null) {
            LinkedHashSet<String> set = new LinkedHashSet<>();
            for (SimpleObjectEntry entry : map) {
                set.add(entry.key);
            }
            this.keySet = set;
        }
        return this.keySet;
    }

    @Override
    public LinkedHashSet<String> keySet() {
        return new LinkedHashSet<>(_keySet());
    }

    @Override
    public List<String> keyList() {
        if (keyList == null) {
            List<String> list = new ArrayList<>(map.size());
            for (SimpleObjectEntry entry : map) {
                list.add(entry.key);
            }
            keyList = list;
        }
        return new ArrayList<>(keyList);
    }

    @Override
    public int size() {
        return map.size();
    }

    @Override
    public boolean containsKey(String key) {
        return _keySet().contains(key);
    }

    private Map<String, JSON.Instance> getFastSingleMap() {
        if (fastSingleMap == null) {
            fastSingleMap = new HashMap<>(map.size());
        }
        return fastSingleMap;
    }

    @Override
    public JSON.Instance get(String key) throws NullPointerException, NoSuchElementException {
        if (key == null)
            throw new NullPointerException();

        Map<String, JSON.Instance> fastMap = getFastSingleMap();
        if (fastMap.containsKey(key)) {
            return fastMap.get(key);
        }
        JSON.Instance inst = null;
        for (SimpleObjectEntry<JSON.Instance> entry : map) {
            if (entry.key.equals(key)) {
                inst = entry.value;
                break;
            }
        }
        if (inst == null)
            throw new NoSuchElementException();
        fastMap.put(key, inst);
        return inst;
    }

    private Map<String, List<JSON.Instance>> getFastMultiMap() {
        if (fastMultiMap == null) {
            fastMultiMap = new HashMap<>();
        }
        return fastMultiMap;
    }

    @Override
    public List<JSON.Instance> getAll(String key) {
        if (key == null)
            throw new NullPointerException();
        if (!_keySet().contains(key)) {
            throw new NoSuchElementException();
        }

        Map<String, List<JSON.Instance>> fastMap = getFastMultiMap();
        if (fastMap.containsKey(key)) {
            return fastMap.get(key);
        }
        List<JSON.Instance> ret = new LinkedList<>();
        for (SimpleObjectEntry<JSON.Instance> entry : map) {
            if (entry.key.equals(key)) {
                ret.add(entry.value);
            }
        }
        ret = Collections.unmodifiableList(ret);
        fastMap.put(key, ret);
        return ret;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof JSON.Object)) return false;
        JSON.Object that = (JSON.Object) o;
        if (!that.keySet().equals(_keySet())) return false;
        for (String key : keySet()) {
            if (!that.get(key).equals(get(key))) return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hash(map);
    }
}

File: base/src/main/kotlin/vjson/simple/SimpleObject.kt
/*
 * The MIT License
 *
 * Copyright 2021 wkgcass (https://github.com/wkgcass)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package vjson.simple

import vjson.JSON
import vjson.JSON.String.Companion.stringify
import vjson.Stringifier
import vjson.parser.TrustedFlag

open class SimpleObject : AbstractSimpleInstance<LinkedHashMap<String, Any?>>, JSON.Object {
  private val map: MutableList<SimpleObjectEntry<JSON.Instance<*>>>
  private var keySet: LinkedHashSet<String>? = null
  private var keyList: List<String>? = null
  private var fastSingleMap: MutableMap<String, JSON.Instance<*>>? = null
  private fun getFastSingleMap(): MutableMap<String, JSON.Instance<*>> {
    if (fastSingleMap == null) {
      fastSingleMap = HashMap(map.size)
    }
    return fastSingleMap!!
  }

  private var fastMultiMap: MutableMap<String, List<JSON.Instance<*>>>? = null
  private fun getFastMultiMap(): MutableMap<String, List<JSON.Instance<*>>> {
    if (fastMultiMap == null) {
      fastMultiMap = HashMap()
    }
    return fastMultiMap!!
  }

  constructor(initMap: Map<String, JSON.Instance<*>>) {
    for ((key, value) in initMap) {
      requireNotNull(key) { "key should not be null" }
      requireNotNull(value) { "value should not be null" }
    }
    map = ArrayList(initMap.size)
    for ((key, value) in initMap) {
      map.add(SimpleObjectEntry(key, value))
    }
  }

  constructor(initMap: List<SimpleObjectEntry<JSON.Instance<*>>>) {
    for (entry in initMap) {
      requireNotNull(entry) { "entry should not be null" }
      // requireNotNull(entry.key) { "key should not be null" }
      // null of the key is tested in the constructor of SimpleObjectEntry
      requireNotNull(entry.value) { "value should not be null" }
    }
    map = ArrayList(initMap)
  }

  protected constructor(initMap: MutableList<SimpleObjectEntry<JSON.Instance<*>>>, flag: TrustedFlag?) {
    if (flag == null) {
      throw UnsupportedOperationException()
    }
    map = initMap
  }

  protected constructor(initMap: MutableList<SimpleObjectEntry<JSON.Instance<*>>>, flag: vjson.util.TrustedFlag?) {
    if (flag == null) {
      throw UnsupportedOperationException()
    }
    map = initMap
  }

  override fun toJavaObject(): LinkedHashMap<String, Any?> {
    return LinkedHashMap(super.toJavaObject())
  }

  override fun _toJavaObject(): LinkedHashMap<String, Any?> {
    val javaObject = LinkedHashMap<String, Any?>()
    for (entry in map) {
      javaObject[entry.key] = entry.value.toJavaObject()
    }
    return javaObject
  }

  override fun stringify(builder: StringBuilder, sfr: Stringifier) {
    sfr.beforeObjectBegin(builder, this)
    builder.append("{")
    sfr.afterObjectBegin(builder, this)
    var isFirst = true
    for (entry in map) {
      if (isFirst) {
        isFirst = false
      } else {
        sfr.beforeObjectComma(builder, this)
        builder.append(",")
        sfr.afterObjectComma(builder, this)
      }
      sfr.beforeObjectKey(builder, this, entry.key)
      builder.append(stringify(entry.key))
      sfr.afterObjectKey(builder, this, entry.key)
      sfr.beforeObjectColon(builder, this)
      builder.append(":")
      sfr.afterObjectColon(builder, this)
      sfr.beforeObjectValue(builder, this, entry.key, entry.value)
      entry.value.stringify(builder, sfr)
      sfr.afterObjectValue(builder, this, entry.key, entry.value)
    }
    sfr.beforeObjectEnd(builder, this)
    builder.append("}")
    sfr.afterObjectEnd(builder, this)
  }

  override fun _toString(): String {
    val sb = StringBuilder()
    sb.append("Object{")
    var isFirst = true
    for (entry in map) {
      if (isFirst) isFirst = false
      else sb.append(", ")
      sb.append(entry.key).append(":").append(entry.value)
    }
    sb.append("}")
    return sb.toString()
  }

  private fun _keySet(): LinkedHashSet<String> {
    if (keySet == null) {
      val set = LinkedHashSet<String>()
      for (entry in map) {
        set.add(entry.key)
      }
      keySet = set
    }
    return keySet!!
  }

  override fun keySet(): LinkedHashSet<String> {
    return LinkedHashSet(_keySet())
  }

  override fun keyList(): List<String> {
    if (keyList == null) {
      val list: MutableList<String> = ArrayList(map.size)
      for (entry in map) {
        list.add(entry.key)
      }
      keyList = list
    }
    return ArrayList(keyList!!)
  }

  override fun size(): Int {
    return map.size
  }

  override fun containsKey(key: String): Boolean {
    return _keySet().contains(key)
  }

  @Throws(NoSuchElementException::class)
  override fun get(key: String): JSON.Instance<*> {
    val fastMap = getFastSingleMap()
    if (fastMap.containsKey(key)) {
      return fastMap[key]!!
    }
    var inst: JSON.Instance<*>? = null
    for (entry in map) {
      if (entry.key == key) {
        inst = entry.value
        break
      }
    }
    if (inst == null) throw NoSuchElementException()
    fastMap[key] = inst
    return inst
  }

  override fun getAll(key: String): List<JSON.Instance<*>> {
    if (!_keySet().contains(key)) {
      throw NoSuchElementException()
    }

    val fastMap = getFastMultiMap()
    if (fastMap.containsKey(key)) {
      return fastMap[key]!!
    }
    val ret: MutableList<JSON.Instance<*>> = ArrayList()
    for (entry in map) {
      if (entry.key == key) {
        ret.add(entry.value)
      }
    }
    val immutableRet: List<JSON.Instance<*>> = ret
    fastMap[key] = immutableRet
    return immutableRet
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other !is JSON.Object) return false
    if (other.keySet() != _keySet()) return false
    for (key in keySet()) {
      if (other[key] != get(key)) return false
    }
    return true
  }

  override fun hashCode(): Int {
    return map.hashCode()
  }
}
